### PR TITLE
feat[next] high-level field storage API

### PR DESCRIPTION
**Pull request by havogt**
Introduce user API to allocate fields in `gt4py.next`.

Summary of main changes:
- Introduce FieldBuffer allocator protocols and implementations
- Introduce the concept of Backend as ProgramExecutor & Allocator
- Replace np_as_located_field with as_field
- Make NdArrayField public
- Fixes for _core.definitions typings
- Fixes and extensions of eve.extended_typing
- Refactor the handling of backends/program processors in the testing infrastructure with string enumerations representing the qualified name of the Python symbol, which can be loaded on demand
- Rename some executor symbols and modules
- Minor style changes to imports and imported symbols to follow coding guidelines.

Open"To Do"s for future PRs:
- Add support for `GTFieldInterface` protocol in cartesian and use it instead of `NextGTDimsInterface` protocol in next.
- Add support for `aligned_index != None` in `FieldBufferAllocator` implementations
- Add support for zero-copy construction of fields in `constructors.as_field()`

### CODE REVIEW

File: src/gt4py/next/common.py [hunk 1]
```diff
@@ -142,7 +142,7 @@ def __str__(self) -> str:
         return f"({self.start}:{self.stop})"
 
 
-RangeLike: TypeAlias = UnitRange | range | tuple[int, int]
+RangeLike: TypeAlias = UnitRange | range | tuple[int, int] | int
```
Comment (havogt) [inline src/gt4py/next/common.py:145]:
```suggestion
RangeLike: TypeAlias = UnitRange | range | tuple[core_defs.INTEGRAL_TYPES, core_defs.INTEGRAL_TYPES] | core_defs.INTEGRAL_TYPES
```

**Review by havogt — COMMENTED** [review]
Just submitting this tiny partial review before it gets lost...

**Review by havogt — COMMENTED** [review]
submitting current state of review. will now look at allocators.py and constructors.py

File: src/gt4py/next/allocators.py [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+    Args: TODO
+        domain:
+        dtype: Data type descriptor as defined in :meth:`BufferAllocator.allocate`.
+
+        aligned_index: N-dimensional index of the first aligned element as defined
+            in :meth:`BufferAllocator.allocate`.
+    """
+    if device is None and allocator is None:
+        raise ValueError("No 'device' or 'allocator' specified")
+    actual_allocator = get_allocator(allocator, None) if allocator is not None else None
```
Comment (havogt) [inline src/gt4py/next/allocators.py:289]: It seems `get_allocator`'s 2nd argument `default` is never used, but it seems it was introduced to simplify the following code...

**Review by havogt — COMMENTED** [review]
I propose we document all TODOs and then merge (after resolving the obvious issues that come/came up in code review and CI).

File: src/gt4py/next/allocators.py (hunk 2 above)
Comment (egparedes) [reply src/gt4py/next/allocators.py]: Fixed

**Comment by egparedes** [comment]
> I propose we document all TODOs 
Added open TODOs to the PR description

File: src/gt4py/next/common.py [hunk 3]
```diff
@@ -397,6 +418,19 @@ def __gt_dims__(self) -> tuple[Dimension, ...]:
         ...
 
 
+class GTFieldInterface(Protocol):
+    """
+    A `GTDimsInterface` is an object providing the `__gt_dims__` property, naming :class:`Field` dimensions.
+
+    The dimension names are objects of type :class:`Dimension`, in contrast to :mod:`gt4py.cartesian`,
+    where the labels are `str` s with implied semantics, see :class:`~gt4py._core.definitions.GTDimsInterface` .
```
[the remark this answers is not in the document]
Comment (egparedes) [reply src/gt4py/next/common.py]: Fixed

File: src/gt4py/next/common.py [hunk 4]
```diff
@@ -397,6 +418,19 @@ def __gt_dims__(self) -> tuple[Dimension, ...]:
         ...
 
 
+class GTFieldInterface(Protocol):
+    """
+    A `GTDimsInterface` is an object providing the `__gt_dims__` property, naming :class:`Field` dimensions.
```
[the remark this answers is not in the document]
Comment (egparedes) [reply src/gt4py/next/common.py]: Fixed

File: src/gt4py/next/program_processors/formatters/pretty_print.py [hunk 5]
```diff
@@ -14,15 +14,23 @@
...
-    pretty = pformat(program)
-    parsed = pparse(pretty)
-    assert parsed == program
+class _RemoveITIRSymTypes(eve.NodeTranslator):
+    def visit_Sym(self, node: itir.Sym) -> itir.Sym:
+        return itir.Sym(id=node.id, dtype=None, kind=None)
+
+
+@ppi.program_formatter
+def format_itir(program: itir.FencilDefinition, *args: Any, **kwargs: Any) -> str:
```
[the remark this answers is not in the document]
Comment (egparedes) [reply src/gt4py/next/program_processors/formatters/pretty_print.py]: Fixed

File: src/gt4py/next/iterator/embedded.py [hunk 6]
```diff
@@ -1012,25 +1012,6 @@ def _shift_field_indices(
     )
 
 
-def np_as_located_field(
```
[the remark this answers is not in the document]
Comment (egparedes) [reply src/gt4py/next/iterator/embedded.py:1015]: Fixed

File: src/gt4py/next/constructors.py [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+            Required if `device` is `None`. If both are valid, `allocator` will be chosen over the default device allocator.
+        device : `Device`
+            The device (CPU, type of accelerator) to optimize the memory layout for.
+            Required if `allocator` is `None` and will cause the default device allocator to be used in that case.
+
+    Returns
+    -------
+        NumPy or CuPy ndarray
+            With uninitialized values, padded and aligned to provide optimal performance for the given `backend` and
+            `aligned_index`
```
[the remark this answers is not in the document]
Comment (egparedes) [reply src/gt4py/next/constructors.py:64]: Fixed

File: src/gt4py/next/constructors.py [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+
+    Raises
+    -------
+        TypeError
+            If arguments of an unexpected type are specified.
+        ValueError
+            If illegal or inconsistent arguments are specified.
```
[the remark this answers is not in the document]
Comment (egparedes) [reply src/gt4py/next/constructors.py:71]: Fixed

File: src/gt4py/next/constructors.py [hunk 9]
```diff
@@ -0,0 +1,353 @@
...
+        device=device,
+    )
+
+    field[...] = field.array_ns.asarray(data)
+
+    return field
+
+
+def as_field_with(
+    domain: common.DomainLike | Sequence[common.Dimension] | eve.NOTHING = eve.NOTHING,
```
[the remark this answers is not in the document]
Comment (egparedes) [reply src/gt4py/next/constructors.py]: Outdated

File: src/gt4py/next/constructors.py [hunk 10]
```diff
@@ -0,0 +1,353 @@
...
+    aligned_index: Optional[Sequence[common.NamedIndex]] = None,
+    allocator: Optional[next_allocators.FieldBufferAllocationTool] = None,
+    device: Optional[core_defs.Device] = None,
+) -> nd_array_field.NdArrayField:
+    """Allocate Field of uninitialized (undefined) values with performance-optimal strides and alignment.
+
+    Parameters
+    ----------
+        domain : `DomainLike`
+            Mapping (or sequence of tuples) of `Dimension` to a range or range-like object.
```
[the remark this answers is not in the document]
Comment (egparedes) [reply src/gt4py/next/constructors.py]: Done

**Review by egparedes — APPROVED** [review]

**Comment by DropD** [comment]
cscs-ci run

**Pull request merged**